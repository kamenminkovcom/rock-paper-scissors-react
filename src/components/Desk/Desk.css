.desk {
    border: black solid 1px;
    width: 40%;
    display: flex;
    flex-direction: column;
}

.desk-half {
    height: 50%;
    width: 100%
}