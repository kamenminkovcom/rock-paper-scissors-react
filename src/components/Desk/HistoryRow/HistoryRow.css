.history-row {
    height: 20%;
}