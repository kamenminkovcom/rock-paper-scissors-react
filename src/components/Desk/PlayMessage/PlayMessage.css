.play-message {
    text-align: center;
    color: #004085;
    background-color: #cce5ff;
    border-color: #b8daff;
    position: relative;
    padding: .75rem 1.25rem;
    margin-bottom: 1rem;
    border: 1px solid transparent;
}