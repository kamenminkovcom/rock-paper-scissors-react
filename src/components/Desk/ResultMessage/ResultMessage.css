.result-message {
    background-image: none;
    border: 2px solid;
    padding: .375rem .75rem;
    font-size: 1rem;
    line-height: 1.5;
    border-radius: .25rem;
    margin: 10%;
    text-align: center;
}