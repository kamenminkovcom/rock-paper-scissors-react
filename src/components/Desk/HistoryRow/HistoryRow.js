import React from 'react';
import PropTypes from 'prop-types';
import clssnames from 'classnames';
import {classNames, possibleResults, handTypes} from '../../../constants/game';

import './HistoryRow.css';

const HistoryRow = ({result, cpuChoice, playerChoice}) => (
    <div className="default-row-container history-row">
        <div className={clssnames(cpuChoice,
            'choice-icon',
            'background-no-repeat',
            'background-contain-size',
            'rotate')}/>
            <span className={clssnames('capitalize', classNames[result])}>{result}</span>
        <div className={clssnames(playerChoice,
            'choice-icon',
            'background-no-repeat',
            'background-contain-size')}/>
    </div>
);

HistoryRow.propTypes = {
    result: PropTypes.oneOf(Object.values(possibleResults)).isRequired,
    cpuChoice: PropTypes.oneOf(Object.values(handTypes)).isRequired,
    playerChoice: PropTypes.oneOf(Object.values(handTypes)).isRequired
};

export default HistoryRow;

