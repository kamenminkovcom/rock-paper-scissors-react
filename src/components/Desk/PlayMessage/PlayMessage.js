import React from 'react';
import withCondition from '../../../HOC/WithCondition';

import './PlayMessage.css';

const PlayMessage = () => (
    <div className="play-message">Please, make your choice.</div>
);

export default withCondition(PlayMessage);