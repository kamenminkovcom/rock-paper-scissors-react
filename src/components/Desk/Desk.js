import React from 'react';
import PropTypes from 'prop-types';
import HistoryRow from './HistoryRow/HistoryRow';
import PlayButton from './PlayButton/PlayButton';
import PlayMessage from './PlayMessage/PlayMessage';
import ResultMessage from './ResultMessage/ResultMessage';
import HistoryHeader from './HistoryHeader/HistoryHeader';
import {handTypes, possibleResults} from '../../constants/game';

import './Desk.css';

const Desk = ({playFunc, isPlaying, history}) => (
    <div className="desk">
        <div className="desk-half">
            <PlayMessage shouldShow={isPlaying}/>
            <PlayButton playFunc={playFunc} shouldShow={!isPlaying}/>
            <ResultMessage message={!!history.length && history[0].result}
                           result={!!history.length && history[0].result}
                           shouldShow={!isPlaying && !!history.length}/>
            <HistoryHeader count={history.length}
                           classes={'history-header'}
                           shouldShow={!!history.length}/>
        </div>
        <div className="desk-half">
            {
                history.map(h => <HistoryRow result={h.result}
                                             cpuChoice={h.cpuChoice}
                                             playerChoice={h.playerChoice}
                                             key={h.id}/>)
            }
        </div>
    </div>
);

Desk.propTypes = {
    playFunc: PropTypes.func.isRequired,
    history: PropTypes.arrayOf(PropTypes.shape({
        cpuChoice: PropTypes.oneOf(Object.values(handTypes)),
        result: PropTypes.oneOf(Object.values(possibleResults)),
        playerChoice: PropTypes.oneOf(Object.values(handTypes)),
        id: PropTypes.string.isRequired
    })).isRequired,
    isPlaying: PropTypes.bool.isRequired
};

export default Desk;