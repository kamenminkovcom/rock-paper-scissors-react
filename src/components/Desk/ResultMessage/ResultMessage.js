import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import withCondition from '../../../HOC/WithCondition';
import {classNames, possibleResults} from '../../../constants/game';

import './ResultMessage.css';

const ResultMessage = ({message, result}) => (
    <div className={classnames(classNames[result],
        'result-message',
        `${classNames[result]}-border`,
        'uppercase')}>
        {message}
    </div>
);

//may we will change the messages, so we are receiving message and result props
//for now they are equal
ResultMessage.propTypes = {
    message: PropTypes.string.isRequired,
    result: PropTypes.oneOf(Object.values(possibleResults))
};

export default withCondition(ResultMessage);