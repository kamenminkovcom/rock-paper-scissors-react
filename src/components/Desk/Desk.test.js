import React from 'react';
import {shallow, mount} from 'enzyme';
import Desk from './Desk';
import HistoryRow from './HistoryRow/HistoryRow';
import PlayMessage from './PlayMessage/PlayMessage';
import ResultMessage from './ResultMessage/ResultMessage';
import {handTypes, possibleResults} from '../../constants/game';
import uid from 'uid';

describe('<Desk/>', () => {
    it('Should render correctly', () => {
        const wrapper = shallow(<Desk history={[]} playFunc={() => {
        }} isPlaying={true}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('Should not render history rows with no history', () => {
        const wrapper = shallow(<Desk history={[]} playFunc={() => {
        }} isPlaying={true}/>);
        expect(wrapper.find(HistoryRow).length === 0).toEqual(true);
    });

    it('Should render HistoryRow instance when have history', () => {
        const game = {
            cpuChoice: handTypes.rock,
            playerChoice: handTypes.paper,
            result: possibleResults.win,
            id: uid()
        };
        const wrapper = shallow(<Desk history={[game]} playFunc={() => {
        }} isPlaying={true}/>);
        expect(wrapper.find(HistoryRow).length === 1).toEqual(true);
    });

    it('Should render multiple HistoryRow instances', () => {
        const historyCount = 10;
        const history = [];
        const choices = Object.values(handTypes);
        for (let i = 1; i <= historyCount; i++) {
            const game = {
                cpuChoice: choices[i % choices.length],
                playerChoice: choices[i % choices.length],
                result: possibleResults.draw,
                id: uid()
            };
            history.push(game);
        }
        const wrapper = shallow(<Desk isPlaying={true} playFunc={() => {}} history={history}/>);
        expect(wrapper.find(HistoryRow).length === historyCount).toEqual(true);
    });

    it('Should not render play massage if isPlaying is false', () => {
        const wrapper = mount(<Desk history={[]} playFunc={() => {
        }} isPlaying={false}/>);
        expect(wrapper.find(PlayMessage).children().length === 0).toEqual(true);
    });

    it('Should render play massage if isPlaying is true', () => {
        const wrapper = mount(<Desk history={[]} playFunc={() => {
        }} isPlaying={true}/>);
        expect(wrapper.find(PlayMessage).children().length > 0).toEqual(true);
    });

    it('Should not render result massage if isPlaying is true', () => {
        const wrapper = mount(<Desk history={[]} playFunc={() => {
        }} isPlaying={true}/>);
        expect(wrapper.find(ResultMessage).children().length === 0).toEqual(true);
    });

    it('Should not render result massage if isPlaying is true and has no history', () => {
        const wrapper = mount(<Desk history={[]} playFunc={() => {
        }} isPlaying={false}/>);
        expect(wrapper.find(ResultMessage).children().length === 0).toEqual(true);
    });

    it('Should render result massage if isPlaying is true and has history', () => {
        const game = {
            cpuChoice: handTypes.rock,
            playerChoice: handTypes.paper,
            result: possibleResults.win,
            id: uid()
        };
        const wrapper = mount(<Desk history={[game]} playFunc={() => {
        }} isPlaying={false}/>);
        expect(wrapper.find(ResultMessage).children().length > 0).toEqual(true);
    });
});