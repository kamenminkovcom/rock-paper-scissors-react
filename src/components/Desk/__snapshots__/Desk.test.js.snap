// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Desk/> Should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Desk
    history={Array []}
    isPlaying={true}
    playFunc={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="desk-half"
>
          <WithCondition
                    shouldShow={true}
          />
          <WithCondition
                    playFunc={[Function]}
                    shouldShow={false}
          />
          <WithCondition
                    message={false}
                    result={false}
                    shouldShow={false}
          />
          <WithCondition
                    classes="history-header"
                    count={0}
                    shouldShow={false}
          />
</div>,
        <div
          className="desk-half"
>
          
</div>,
      ],
      "className": "desk",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <WithCondition
              shouldShow={true}
/>,
            <WithCondition
              playFunc={[Function]}
              shouldShow={false}
/>,
            <WithCondition
              message={false}
              result={false}
              shouldShow={false}
/>,
            <WithCondition
              classes="history-header"
              count={0}
              shouldShow={false}
/>,
          ],
          "className": "desk-half",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "shouldShow": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "playFunc": [Function],
              "shouldShow": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "message": false,
              "result": false,
              "shouldShow": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "classes": "history-header",
              "count": 0,
              "shouldShow": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "desk-half",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="desk-half"
>
            <WithCondition
                        shouldShow={true}
            />
            <WithCondition
                        playFunc={[Function]}
                        shouldShow={false}
            />
            <WithCondition
                        message={false}
                        result={false}
                        shouldShow={false}
            />
            <WithCondition
                        classes="history-header"
                        count={0}
                        shouldShow={false}
            />
</div>,
          <div
            className="desk-half"
>
            
</div>,
        ],
        "className": "desk",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithCondition
                shouldShow={true}
/>,
              <WithCondition
                playFunc={[Function]}
                shouldShow={false}
/>,
              <WithCondition
                message={false}
                result={false}
                shouldShow={false}
/>,
              <WithCondition
                classes="history-header"
                count={0}
                shouldShow={false}
/>,
            ],
            "className": "desk-half",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "shouldShow": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "playFunc": [Function],
                "shouldShow": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "message": false,
                "result": false,
                "shouldShow": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "classes": "history-header",
                "count": 0,
                "shouldShow": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "desk-half",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
